java/vector/src/main/java/org/apache/arrow/vector/complex/RunEndEncodedVector.java

```diff
@@ -312,8 +302,142 @@ public TransferPair getTransferPair(Field field, BufferAllocator allocator, Call
+    }
+
+    public TransferImpl(Field field, BufferAllocator allocator, CallBack callBack) {
+      this(new RunEndEncodedVector(field, allocator, callBack));
+    }
+
+    public TransferImpl(RunEndEncodedVector to) {
+      this.to = to;
+      if (to.getRunEndsVector() instanceof ZeroVector) {
+        to.initializeChildrenFromFields(field.getChildren());
+      }
+      reeTransferPair = getRunEndsVector().makeTransferPair(to.getRunEndsVector());
+      dataTransferPair = getValuesVector().makeTransferPair(to.getValuesVector());
+    }
+
+    /**
+     * Transfer the vector data to another vector. The memory associated with this vector is
+     * transferred to the allocator of target vector for accounting and management purposes.
+     */
+    @Override
+    public void transfer() {
+      to.clear();
+      dataTransferPair.transfer();
+      reeTransferPair.transfer();
+      if (valueCount > 0) {
+        to.setValueCount(valueCount);
+      }
+      clear();
+    }
+
+    /**
+     * Slice this vector at the desired index and length, then transfer the corresponding data to the
+     * target vector.
+     *
+     * @param startIndex start position of the split in source vector.
+     * @param length length of the split.
+     */
+    @Override
+    public void splitAndTransfer(int startIndex, int length) {
+      ValueVector toDataVector = dataTransferPair.getTo();
+      ValueVector toRunEndVector = reeTransferPair.getTo();
+
+      toDataVector.clear();
+      toRunEndVector.clear();
+
+      int endIndex = startIndex + length;
+      int physicalStartIndex = getPhysicalIndex(startIndex);
+      int physicalEndIndex = getPhysicalIndex(endIndex);
+      int physicalLength = physicalEndIndex - physicalStartIndex + 1;
+      dataTransferPair.splitAndTransfer(physicalStartIndex, physicalLength);
+      if (startIndex == 0) {
+        reeTransferPair.splitAndTransfer(physicalStartIndex, physicalLength);
+      } else {
+        shiftRunEndVector(
+            toRunEndVector,
+            startIndex,
+            length,
+            physicalStartIndex,
+            physicalEndIndex,
+            physicalLength);
+      }
+      getTo().setValueCount(length);
+    }
+
+    private void shiftRunEndVector(
+        ValueVector toRunEndVector,
+        int startIndex,
+        int length,
+        int physicalStartIndex,
+        int physicalEndIndex,
+        int physicalLength) {
+      toRunEndVector.setValueCount(physicalLength);
+      toRunEndVector.getValidityBuffer().setOne(0, toRunEndVector.getValidityBuffer().capacity());
+      ArrowBuf fromRunEndBuffer = runEndsVector.getDataBuffer();
+      ArrowBuf toRunEndBuffer = toRunEndVector.getDataBuffer();
+      int physicalLastIndex = physicalLength - 1;
+      if (toRunEndVector instanceof SmallIntVector) {
+        byte typeWidth = SmallIntVector.TYPE_WIDTH;
+        for (int i = 0; i < physicalLastIndex; i++) {
```

Comment:
Shouldn't the loop go from `[0, physicalLength)` or `[physicalStartIndex, physicalLastIndex)`?

Reply:
it through [0, physicalLength - 1), and handle `physicalLength - 1` separately

Reply:
Right, but this loop is from `[0, physicalLastIndex)` - aren't we "crossing indices" here?

Reply:
physicalLastIndex = physicalLength - 1

Reply:
yeah, the naming is kind of confusing, I renamed them, does it make sense for you now?